### PR TITLE
Add AST rewriter for label filter pushdown

### LINKED ISSUE #513: Entity variables in planners have the wrong CypherType
The AST removes the labels from the pattern and moves them into to `WHERE` clauses. They are later added back in the logical optimizer step. However, typing of the pattern and thus typing of the entity variables already happens at IR planning time. The variables and their CypherTypes are then used as input for subsequent planning phases. This leads to a mismatch of the variables stored in the RecordHeader and the variables references in the operators

### PR DESCRIPTION
**Pull request by DarthMax**
Fixes #513 
Co-authored-by: Sören Reichardt <soren.reichardt@neotechnology.com>

### CODE REVIEW

File: okapi-ir/src/test/scala/org/opencypher/okapi/ir/impl/parse/rewriter/pushLabelsIntoScansTest.scala [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      """MATCH (n:Foo)
+        |WHERE n:Bar
+        |RETURN n
+      """.stripMargin,
+      """MATCH (n:Foo:Bar)
+        |RETURN n
+      """.stripMargin)
+  }
+
+  test("push complex where predicate labels into pattern") {
```
Comment (Mats-SX) [inline okapi-ir/src/test/scala/org/opencypher/okapi/ir/impl/parse/rewriter/pushLabelsIntoScansTest.scala]: Let's maybe add a test where the `HasLabel` is composed with another type of predicate as well:

```
MATCH (n) WHERE n.p = 5 AND n:Foo
```

File: okapi-logical/src/test/scala/org/opencypher/okapi/logical/impl/logical/LogicalPlannerTest.scala [hunk 2]
```diff
@@ -108,150 +108,140 @@ class LogicalPlannerTest extends BaseTestSuite with IrConstruction {
...
+//          HasLabel(Var("a")(CTNode), Label("Administrator"))(CTBoolean),
+//          HasLabel(Var("g")(CTNode), Label("Group"))(CTBoolean),
+//          Equals(Property(Var("g")(CTNode(Set("Group"))), PropertyKey("name"))(CTString), Param("foo")(CTString))(
+//            CTBoolean)
+//        )
+//      )
+//    )
+//
+//    result should equalWithoutResult(expected)
+//  }
```
Comment (Mats-SX) [inline okapi-logical/src/test/scala/org/opencypher/okapi/logical/impl/logical/LogicalPlannerTest.scala]: Either delete or uncomment and fix this. I'm not really comfortable just deleting it as it is the only tests we have that test the pipeline at this level. It's pretty tedious to write though...

File: okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalOptimizer.scala [hunk 3]
```diff
@@ -26,35 +26,24 @@
...
       discardScansForNonexistentLabels,
       replaceCartesianWithValueJoin)
-    optimizationRules.foldLeft(input) {
+      optimizationRules.foldLeft(input) {
       // TODO: Evaluate if multiple rewriters could be fused
       case (tree: LogicalOperator, optimizationRule) => BottomUp[LogicalOperator](optimizationRule).transform(tree)
     }
   }
 
-  def labelsForVariables(op: LogicalOperator): Map[Var, Set[String]] = {
```
Comment (Mats-SX) [inline okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalOptimizer.scala]: Nice replacement

File: okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalOperatorProducer.scala [hunk 4]
```diff
@@ -82,7 +82,12 @@ class LogicalOperatorProducer {
   }
 
   def planNodeScan(node: IRField, prev: LogicalOperator): NodeScan = {
-    NodeScan(node, prev, prev.solved.withField(node))
+    val solvedWithField = prev.solved.withField(node)
+    val solvedWithPredicates = node.cypherType match {
+      case CTNode(labels, _) => solvedWithField.withPredicates(labels.map(l => HasLabel(node, Label(l))(CTBoolean)).toSeq:_*)
+      case _ => solvedWithField
```
Comment (Mats-SX) [inline okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalOperatorProducer.scala]: Isn't this some kind of error?

File: okapi-logical/src/test/scala/org/opencypher/okapi/logical/impl/logical/LogicalPlannerTest.scala (hunk 2 above)
Comment (DarthMax) [reply okapi-logical/src/test/scala/org/opencypher/okapi/logical/impl/logical/LogicalPlannerTest.scala]: We have fixed them for now. However, we should think of a better way to do these tests

File: okapi-ir/src/test/scala/org/opencypher/okapi/ir/impl/parse/rewriter/pushLabelsIntoScansTest.scala (hunk 1 above)
Comment (DarthMax) [reply okapi-ir/src/test/scala/org/opencypher/okapi/ir/impl/parse/rewriter/pushLabelsIntoScansTest.scala]: :+1:

File: okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalOperatorProducer.scala (hunk 4 above)
Comment (DarthMax) [reply okapi-logical/src/main/scala/org/opencypher/okapi/logical/impl/LogicalOperatorProducer.scala]: Not if we assume that the CypherType will always be CTNode in which case the other case just covers the case without labels

**Pull request merged**